File: app/src/main/java/de/andicodes/vergissnix/ui/main/EditTaskViewModel.java
package de.andicodes.vergissnix.ui.main;

import android.app.Application;
import android.content.Context;

import java.time.LocalDateTime;
import java.time.ZonedDateTime;
import java.util.TimeZone;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;
import de.andicodes.vergissnix.NotificationBroadcastReceiver;
import de.andicodes.vergissnix.data.AppDatabase;
import de.andicodes.vergissnix.data.Task;
import de.andicodes.vergissnix.data.TaskDao;

public class EditTaskViewModel extends AndroidViewModel {
    private final MutableLiveData<Task> task = new MutableLiveData<>(new Task());
    private final MutableLiveData<String> text = new MutableLiveData<>();
    private final MutableLiveData<LocalDateTime> customDatetime = new MutableLiveData<>();
    private final MutableLiveData<LocalDateTime> recommendationDatetime = new MutableLiveData<>();
    private final TaskDao taskDao;

    private final Observer<Task> taskObserver = task -> {
        if (task != null) {
            text.setValue(task.getText());
            if (task.getTime() != null) {
                setCustomDatetime(task.getTime().toLocalDateTime());
            } else {
                setCustomDatetime(null);
            }
        } else {
            text.setValue(null);
            setCustomDatetime(null);
        }
    };

    public EditTaskViewModel(@NonNull Application application) {
        super(application);
        this.taskDao = AppDatabase.getDatabase(application).taskDao();
        task.observeForever(taskObserver);
    }

    @Override
    protected void onCleared() {
        task.removeObserver(taskObserver);
        super.onCleared();
    }

    public Task getTask() {
        Task task = this.task.getValue();
        if (task == null) {
            task = new Task();
        }
        task.setText(text.getValue());
        if (customDatetime.getValue() != null) {
            task.setTime(ZonedDateTime.of(customDatetime.getValue(), TimeZone.getDefault().toZoneId()));
        }
        if (recommendationDatetime.getValue() != null) {
            task.setTime(ZonedDateTime.of(recommendationDatetime.getValue(), TimeZone.getDefault().toZoneId()));
        }
        return task;
    }

    public void setTask(Task task) {
        this.task.setValue(task);
    }

    public void setCustomDatetime(LocalDateTime customDatetime) {
        this.recommendationDatetime.setValue(null);
        this.customDatetime.setValue(customDatetime);
    }

    public LiveData<LocalDateTime> getCustomDatetime() {
        return customDatetime;
    }

    public void setRecommendationDatetime(LocalDateTime datetime) {
        this.recommendationDatetime.setValue(datetime);
        this.customDatetime.setValue(null);
    }

    public LiveData<LocalDateTime> getRecommendationDatetime() {
        return recommendationDatetime;
    }

    public MutableLiveData<String> getText() {
        return text;
    }

    public void saveCurrentTask(Context context) {
        AppDatabase.databaseWriteExecutor.execute(() -> {
            Task result = taskDao.saveTask(getTask());
            NotificationBroadcastReceiver.setNotificationAlarm(context, result);
            task.postValue(null);
        });
    }
}

File: app/src/main/java/de/andicodes/vergissnix/ui/main/EditTaskViewModel.kt
package de.andicodes.vergissnix.ui.main

import android.app.Application
import android.content.Context
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.Observer
import androidx.lifecycle.viewModelScope
import de.andicodes.vergissnix.NotificationBroadcastReceiver
import de.andicodes.vergissnix.data.AppDatabase.Companion.getDatabase
import de.andicodes.vergissnix.data.Task
import de.andicodes.vergissnix.data.TaskDao
import kotlinx.coroutines.flow.FlowCollector
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.time.LocalTime
import java.time.ZonedDateTime
import java.util.TimeZone

@ExperimentalMaterial3Api
class EditTaskViewModel(application: Application) : AndroidViewModel(application) {
    private val _task = MutableStateFlow<Task?>(Task())
    private val _text = MutableStateFlow<String?>(null)
    private val _time = MutableStateFlow<LocalTime?>(null)
    private val _date = MutableStateFlow<LocalDate?>(null)

    private val taskDao: TaskDao = getDatabase(application)!!.taskDao()

    private val taskObserver = FlowCollector { task: Task? ->
        if (task != null) {
            _text.value = task.text
            _time.value = task.time?.toLocalTime()
            _date.value = task.time?.toLocalDate()
        } else {
            _text.value = null
            _time.value = null
            _date.value = null
        }
    }

    init {
        viewModelScope.launch {
            _task.collect(taskObserver)
        }
    }

    fun loadTaskById(taskId: Long) {
        viewModelScope.launch {
            taskDao.getTask(taskId).let { _task.value = it }
        }
    }

    private fun getTaskToSave(): Task {
        var task = _task.value
        if (task == null) {
            task = Task()
        }
        task.text = _text.value
        task.time = ZonedDateTime.of(_date.value, _time.value, TimeZone.getDefault().toZoneId())
        return task
    }

    fun saveTask(context: Context) {
        viewModelScope.launch {
            val result = taskDao.saveTask(getTaskToSave())
            NotificationBroadcastReceiver.setNotificationAlarm(context, result)
            _task.value = null
        }
    }

    val time = _time.asStateFlow()
    fun setTime(time: LocalTime) {
        this._time.value = time
    }

    val date = _date.asStateFlow()
    fun setDate(date: LocalDate) {
        this._date.value = date
    }

    val originalTask = _task.asStateFlow()

    val text = _text.asStateFlow()
    fun setText(text: String) {
        this._text.value = text
    }
}